apps/service_gather/lib/gather/extraction.ex

```diff
@@ -28,28 +28,56 @@ defmodule Gather.Extraction do
   @dialyzer {:nowarn_function, handle_continue: 2}
   @impl GenServer
   def handle_continue(:extract, %{extract: extract} = state) do
-    Logger.debug(fn -> "#{__MODULE__}: Started extraction: #{inspect(extract)}" end)
-
     case extract(extract) do
-      {:ok, _destination_pid} ->
-        Logger.debug(fn -> "#{__MODULE__}: Extraction Completed: #{inspect(extract)}" end)
-        Brook.Event.send(Gather.Application.instance(), extract_end(), "gather", extract)
-        {:stop, :normal, state}
+      {:ok, pid} ->
+        Logger.debug(fn -> "#{__MODULE__}: Started extraction: #{inspect(extract)}" end)
+        {:noreply, Map.put(state, :destination_pid, pid)}
 
       {:error, reason} ->
         Logger.warn("#{__MODULE__}: Extraction Stopping: #{inspect(extract)}")
         {:stop, reason, state}
     end
   end
 
+  @impl GenServer
+  def handle_info(:extract_complete, %{extract: extract, destination_pid: pid} = state) do
+    Destination.stop(extract.destination, pid)
+
+    Logger.debug(fn -> "#{__MODULE__}: Extraction Completed: #{inspect(extract)}" end)
+    Brook.Event.send(Gather.Application.instance(), extract_end(), "gather", extract)
+
+    {:stop, :normal, state}
+  end
+
+  @impl GenServer
+  def handle_info({:extract_failed, reason}, %{extract: extract, destination_pid: pid} = state) do
+    Destination.stop(extract.destination, pid)
+
+    Logger.warn("#{__MODULE__}: Extraction Stopping: #{inspect(extract)}")
+    {:stop, reason, state}
+  end
+
+  @impl GenServer
+  def handle_info(msg, state) do
+    Logger.info(fn -> "Received message : #{inspect(msg)}" end)
+    {:noreply, state}
+  end
+
   @retry with: exponential_backoff(@initial_delay) |> take(@max_tries)
   defp extract(extract) do
     with {:ok, pid} <- start_destination(extract),
-         :ok <- do_extract(pid, extract) do
+         :ok <- start_source(extract, pid) do
       {:ok, pid}
     end
   end
 
+  def start_source(extract, destination_pid) do
+    {:ok, _source_pid} =
```

Comment:
I think we need to keep a handle on this pid. While it absolutely will not be useful for HTTP-related sources, we'll probably want it for Kafka sources. If we want to delete the dataset or shut it down later, we'll want to stop those source processes.

Reply:
I'm not suggesting you need to worry about the delete/shutdown scenario in this PR, just that we should track it in state.

Reply:
Addressed; let us know what you think.